libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookClientWrapper.cs

```diff
@@ -61,6 +64,78 @@ public async Task<HttpResponseMessage> CallAPIAsync(string path, FacebookMessage
             }
         }
 
+        /// <summary>
+        /// Get a Facebook API client with the correct credentials based on the page identified in the incoming activity.
+        /// This is used by many internal functions to get access to the Facebook API, and is exposed as `bot.api` on any BotWorker instances passed into Botkit handler functions.
+        /// </summary>
+        /// <param name="activity">An incoming message activity.</param>
+        /// <returns>A Facebook API client.</returns>
+        public async Task<FacebookClientWrapper> GetAPIAsync(Activity activity)
+        {
+            if (!string.IsNullOrWhiteSpace(Options.AccessToken))
+            {
+                return new FacebookClientWrapper(new FacebookAdapterOptions(Options.VerifyToken, Options.AppSecret, Options.AccessToken));
+            }
+            else
+            {
+                if (!string.IsNullOrWhiteSpace(activity.Recipient?.Id))
+                {
+                    var pageId = activity.Recipient.Id;
+
+                    if ((activity.ChannelData as dynamic)?.message != null && (activity.ChannelData as dynamic)?.message.is_echo)
+                    {
+                        pageId = activity.From.Id;
+                    }
+
+                    var token = await Options.GetAccessTokenForPageAsync(pageId).ConfigureAwait(false);
+
+                    if (string.IsNullOrWhiteSpace(token))
+                    {
+                        // error: missing credentials
```

Comment:
Should this throw an error?

Reply:
should probably throw an exception
                    `if (string.IsNullOrWhiteSpace(token))
                    {
                        throw new ArgumentException(nameof(token));
                    }`

Reply:
Yes, here you should probably throw an exception.